File: codility/js/11-CountSemiprimes.js
function solution(N, P, Q) {
  // Record number of primes for each value 1...N
  const facts = {}
  const semis = { 0: 0, 1: 0 }
  let i = 2
  while (i <= N) {
    // Record first prime factors for all values
    if (!facts[i]) {
      let k = i * i
      while (k <= N) {
        if (!facts[k]) facts[k] = i
        k += i
      }
    }

    const div = i / facts[i]
    const isDivPrime = div > 0 && !facts[div] // If never recorded in a map, it is not a prime
    const prevSemis = semis[i - 1] || 0
    semis[i] = prevSemis + (isDivPrime ? 1 : 0)

    i++
  }

  return P.map((_, idx) => {
    return (semis[Q[idx]] || 0) - (semis[P[idx] - 1] || 0) // Semis on P[idx] inclusive ( add  - 1 )
  })
}

// TEST
const p = [1, 4, 16]
const q = [26, 10, 20]
const n = Math.max(...q)
const f = solution(n, p, q)

File: codility/py/11-count_semiprimes.py
# https://app.codility.com/programmers/lessons/11-sieve_of_eratosthenes/count_semiprimes/
# A semiprime is a natural number that is the product of two (not necessarily distinct) prime numbers. The first few semiprimes are 4, 6, 9, 10, 14, 15, 21, 22, 25, 26.


def solution(N, P, Q):
    # Find smallest factors for 1...N. Used to determine semiprimes
    factors = [0] * (N + 1)

    # Keep track of total semiprimes found
    # Record the count for each number from Q and P(-1)
    semi_count = 0  # Keep track of total semis found
    endpoints = {}
    for p, q in zip(P, Q):
        endpoints[p - 1] = 0
        endpoints[q] = 0

    for n in range(2, N + 1):
        # Record smallest factors
        if factors[n] == 0:
            k = n * n
            while k < len(factors):
                if factors[k] == 0:
                    factors[k] = n
                k += n

        # Update total semis count
        semi_count += 1 if factors[n] != 0 and factors[n // factors[n]] == 0 else 0

        # If n is also an endpoint from P and Q, record the current count
        if n in endpoints:
            endpoints[n] = semi_count

    return [endpoints[q] - endpoints.get(p - 1, 0) for p, q in zip(P, Q)]


assert solution(26, [1, 4, 16], [26, 10, 20]) == [10, 4, 0]
